rosidl_runtime_py/rosidl_runtime_py/convert.py

```diff
@@ -85,74 +101,107 @@ def to_string(val):
-        result += to_string(value)
+
+        result += to_string(value, field_type, has_maximum_size)
     return result
 
 
 # Convert a msg to an OrderedDict. We do this instead of implementing a generic __dict__() method
 # in the msg because we want to preserve order of fields from the .msg file(s).
-def message_to_ordereddict(msg: Any, truncate_length: int = None) -> OrderedDict:
+def message_to_ordereddict(
+        msg: Any,
+        truncate_length: int = None,
+        no_arr: bool = False,
+        no_str: bool = False) -> OrderedDict:
     """
     Convert a ROS message to an OrderedDict.
 
     :param msg: The ROS message to convert.
     :param truncate_length: Truncate values for all message fields to this length.
         This does not truncate the list of fields (ie. the dictionary keys).
+    :param no_arr: Exclude array fields of the message
+    :param no_str: Exclude string fields of the message
     :returns: An OrderedDict where the keys are the ROS message fields and the values are
         set to the values of the input message.
     """
     d = OrderedDict()
+    slot_types = get_message_slot_types(msg)
+
     # We rely on __slots__ retaining the order of the fields in the .msg file.
     for field_name in msg.__slots__:
         value = getattr(msg, field_name, None)
-        value = _convert_value(value, truncate_length=truncate_length)
+        field_type = msg._fields_and_field_types[field_name[1:]]
+        has_maximum_size = getattr(slot_types[field_name[1:]], 'has_maximum_size', None)
+
+        value = _convert_value(
+            value, field_type, has_maximum_size,
+            truncate_length=truncate_length, no_arr=no_arr, no_str=no_str)
         # Remove leading underscore from field name
         d[field_name[1:]] = value
     return d
 
 
-def _convert_value(value, truncate_length=None):
+def _convert_value(value, field_type=None, has_maximum_size=None,
+                   truncate_length=None, no_arr=False, no_str=False):
+
     if isinstance(value, bytes):
         if truncate_length is not None and len(value) > truncate_length:
             value = ''.join([chr(c) for c in value[:truncate_length]]) + '...'
         else:
             value = ''.join([chr(c) for c in value])
     elif isinstance(value, str):
-        if truncate_length is not None and len(value) > truncate_length:
+        if no_str is True:
+            value = '<string length: <{0}>>'.format(len(value))
+        elif truncate_length is not None and len(value) > truncate_length:
             value = value[:truncate_length] + '...'
     elif isinstance(value, (list, tuple, array.array, numpy.ndarray)):
         # Since arrays and ndarrays can't contain mixed types convert to list
         typename = tuple if isinstance(value, tuple) else list
-        if truncate_length is not None and len(value) > truncate_length:
+        if no_arr is True and field_type is not None:
+            if(callable(has_maximum_size) and has_maximum_size()):
```

Comment:
@vinnamkim meta: I think you can derive both `field_type` and `maximum_size` from the value type, but that doesn't do much to the actual functionality. @dirk-thomas I'd like to know what you think.

Reply:
I guess `field_type` is well defined in `_fields_and_field_types` of the parent of the `value`. It would be more complicated to derive `field_type` from the `value`.